src/Ledger/Ledger/Properties.agda

```diff
@@ -82,29 +87,536 @@ instance
+    updateVotesOnly s [] = s
+    updateVotesOnly s (inj₁ v ∷ vps) = updateVotesOnly (updateVote s v) vps
+    updateVotesOnly s (inj₂ _ ∷ vps) = updateVotesOnly s vps
+
+    votesOnly-empty : (vps : List (GovVote ⊎ GovProposal)) → updateVotesOnly [] vps ≡ []
+    votesOnly-empty [] = refl
+    votesOnly-empty (inj₁ v ∷ vps) = votesOnly-empty vps
+    votesOnly-empty (inj₂ p ∷ vps) = votesOnly-empty vps
+
+    mkAction : GovProposal → ℕ → GovActionID × GovActionState
+    mkAction p n = let open GovProposal p in
+      ActionId×ActionState (PParams.govActionLifetime pp +ᵉ GovEnv.epoch gΓ)
+                           (GovEnv.txid gΓ , n) returnAddr action prevAction
+
+    propToState : GovState → GovProposal → ℕ → GovState
+    propToState s p n = s ∷ʳ mkAction p n
+
+    propToState-++ : (gs gs' : GovState) (p : GovProposal) (n : ℕ)
+                    → propToState (gs ++ gs') p n ≡ gs ++ propToState gs' p n
+    propToState-++ gs gs' p n = let open ≡-Reasoning in
+      begin
+      propToState (gs ++ gs') p n  ≡⟨ ++-assoc gs gs' [ mkAction p n ] ⟩
+      gs ++ gs' ∷ʳ mkAction p n    ≡⟨ cong (gs ++_) refl ⟩
+      gs ++ propToState gs' p n    ∎
+
+    updateGovStates : GovState → List (GovVote ⊎ GovProposal) → GovState
+    updateGovStates s [] = s
+    updateGovStates s (inj₁ v ∷ vps) = updateGovStates (updateVote s v) vps
+    updateGovStates s (inj₂ p ∷ vps) = updateGovStates (propToState s p (length vps)) vps
+
+    STS→updateGovSt≡ : {govSt govSt' : GovState} (vps : List (GovVote ⊎ GovProposal))
+                       → (gΓ ⊢ govSt ⇀⦇ vps ,GOV⦈ govSt')
+                       → govSt' ≡ updateGovStates govSt vps
+
+    STS→updateGovSt≡ [] (BS-base Id-nop) = refl
+    STS→updateGovSt≡ (inj₁ v ∷ vps) (BS-ind (_⊢_⇀⦇_,GOV'⦈_.GOV-Vote x) x₁) = STS→updateGovSt≡ vps x₁
+    STS→updateGovSt≡ (inj₂ p ∷ vps) (BS-ind (_⊢_⇀⦇_,GOV'⦈_.GOV-Propose x) x₁) = STS→updateGovSt≡ vps x₁
+
+    govUpdate_withIsValid≡_ : LState → Bool → GovState
+    govUpdate s withIsValid≡ true = updateGovStates govSt (txgov txb)
+    govUpdate s withIsValid≡ false = govSt
+
+
+    STS→GovSt≡' : ∀ {s' : LState} → Γ ⊢ s ⇀⦇ tx ,LEDGER⦈ s'
+                 → LState.govSt s' ≡ govUpdate s withIsValid≡ isValid
+
+    STS→GovSt≡' {⟦ _ , govSt' , _ ⟧ˡ} (LEDGER-V x) = let open ≡-Reasoning in
+      begin
+      govSt'                             ≡⟨ STS→updateGovSt≡ (txgov txb) (proj₂ (proj₂ (proj₂ x))) ⟩
+      updateGovStates govSt (txgov txb)  ≡˘⟨ cong (govUpdate s withIsValid≡_) (proj₁ x) ⟩
+      govUpdate s withIsValid≡ isValid   ∎
+
+    STS→GovSt≡' {⟦ _ , .govSt , _ ⟧ˡ} (LEDGER-I x) = sym (cong (govUpdate s withIsValid≡_) (proj₁ x))
+
+    STS→GovSt≡ : ∀ {s' : LState} → Γ ⊢ s ⇀⦇ tx ,LEDGER⦈ s'
+                 → isValid ≡ true
+                 → LState.govSt s' ≡ updateGovStates govSt (txgov txb)
+    STS→GovSt≡ utxo-sts txvalid = trans (STS→GovSt≡' utxo-sts) (cong (govUpdate s withIsValid≡_) txvalid)
```

Comment:
You can prove this more easily directly:
```suggestion
    STS→GovSt≡ (LEDGER-V x) refl = STS→updateGovSt≡ (txgov txb) (proj₂ (proj₂ (proj₂ x)))
```
Then, `govUpdate_withIsValid≡_` and `STS→GovSt≡'` become unused.

Reply:
Nice! 👍🏼